### ISSUE TITLE
the code implementation doesn't seem to have Hyperbolic TransE and hyperbolic operations

### ISSUE THREAD

**Issue by cs-maillist** [DEEP-PolyU/HamQA_WWW23#2]
Hello authors,
I've been working on this paper and code recently, and found some problems, and came to ask for advice:

1. In the paper, it is joint learning, and the final loss is Eq. 12. In the code, HamQA.py only uses cross-entropy loss, line 232.
2. Don't seem to find an implementation of the hyperbolic TransE KG embedding part in the code.
3. The model is LM_HamQA, its decoder is HamQA --> HamQA_Messege_Passing --> HamLayer, and doesn't see the use of hyperbolic operations. modeling/modeling_hamqa.py defines HyperGAT and other hyperbolic layer, doesn't seem to be used?

I feel that this paper is still useful, I hope the author can help to clear up the confusion, thanks!

**Comment by Qing145** [comment]
Dear cs-maillist, thanks a lot for your attention for your paper! We summarize the functions used for hyperbolic KG embedding in file 'utils/hyperbolic_utils.py', and we omit the training details in this repository since the whole KG is pretrained with the hyperbolic KG embedding methods, and we only kept the pretrained embedding in file 'data/cpnet/tzw.ent.npy' that used at Line 37&43 in file '/utils/parser_utils.py'. As for the jointly learning framework, we adopt the cross-entropy loss as well as the trainable embedding table to make the embeddings of entities related to the same questions being close to each other in hyperbolic vector space. Since I don't have the right to update this repo, please refer to my own repo for latest updates (https://github.com/Qing145/HamQA.git). I'll complete the missing pretrain part to make the implementation more clear. Thanks.
